Core/include/Acts/Utilities/VectorHelpers.hpp

```diff
@@ -119,7 +120,11 @@ double eta(const Eigen::MatrixBase<Derived>& v) noexcept {
     assert(v.rows() == 3 && "Eta function not valid for non-3D vectors.");
   }
 
-  return std::atanh(v[2] / v.norm());
+  if (v[0] == 0. && v[1] == 0.) {
+    return std::copysign(std::numeric_limits<double>::infinity(), v[2]);
+  } else {
+    return std::asinh(v[2] / perp(v));
+  }
```

Comment:
one optimization opportunity might be `v[0] * v[1] == 1` but I would not change it without benchmarking

other than that I dont see a way around the branch (which should be fine on a CPU I think)

Reply:
oh that would represent an `||` ... dont mind me :smile:

Reply:
Hi, I can't think of any way to remove the branching. I'm wondering if something like
```c++
double pT=perp(v);
if(pT==0) { ... }
else { 
  return std::asinh(v[2] / pT);
}
``` 
would be faster.

Reply:
Ok, after playing for a while, I can only conclude that the timing precision I have can't reliably distinguish between versions with `==`+`&&` or `!=`+`||`, with temporary variable `pT` or without, and even with calling `std::isnormal` instead of `!=`.
Thus, I don't think there's anything we can improve here.